rlapy/comps/evd.py

```diff
@@ -0,0 +1,432 @@
+            the matrix A
+        """
+        raise NotImplementedError()
+
+
+class EVD1(EVDecomposer):
+
+    TOL_CONTROL = 'unknown'  # depends on implementation of rangefinder
+
+    def __init__(self, rf: RangeFinder):
+        """
+        Parameters
+        ----------
+        rf : RangeFinder
+        """
+        self.rangefinder = rf
+
+    def __call__(self, A, k, epsilon, s, rng):
+        """
+        Rely on a rangefinder to obtain the matrix Q for the decomposition
+        A \approx Q B. Once we have Q, we construct B = Q.T @ A and return
+        (Q, B). This function is agnostic to the implementation of the
+        rangefinder: it might build a rank-k matrix Q all at once or construct
+        successively larger matrices Q by an iterative process. We make no
+        assumptions on the rangefinder's termination criteria beyond those
+        listed below.
+
+        Parameters
+        ----------
+        num_passes : int
+            Total number of passes over A. We require num_passes >= 2.
+
+        A : Union[ndarray, spmatrix, LinearOperator]
+            Data matrix to approximate. A must be an n × n Hermitian matrix.
+
+        k : int
+            Target rank for the approximation of A: 0 < k < min(A.shape).
+            This parameter includes any oversampling. For example, if you
+            want to be near the optimal (Eckhart-Young) error for a rank 20
+            approximation of A, then you might want to set k=25.
+            
+        epsilon : float
+            Target accuracy for the oversampled approximation of A: 0 < epsilon < np.inf.
+            This parameter inherits from the QBDecomposer or RangeFinder class.
+            
+        s  : int
+            Auxiliary parameter for the QBDecomposer or RangeFinder.
+
+        rng : Union[None, int, SeedSequence, BitGenerator, Generator]
+            Determines the numpy Generator object that manages randomness
+            in this function call.
+
+        Returns
+        -------
+        V : ndarray
+            Has shape (A.shape[0], k). Columns are orthonormal.
+
+        lambda_matrix : ndarray
+            Has shape (k,k). lambda_matrix is a diagonal matrix storing the eigenvalues of 
+            the matrix A
+
+        """
+        assert k > 0
+        assert k <= min(A.shape)
+        if not np.isnan(epsilon):
+            assert tol >= 0
+            assert tol < np.inf
+        rng = np.random.default_rng(rng)
+        Q = self.rangefinder(A, k, tol, rng)
+        C = B @ Q
+        # d = number of columns in Q, d ≤ k + s
+        d = Q.shape[1]
+        if d > k+s:
+            msg = """
+            This implementation the dimension of Q matrix <= k + s.
+            """ 
+            raise RuntimeError(msg)
+
+        U, lambda_matrix, Vh = la.svd(C)
+        # Full d × d Hermitian eigendecomposition for the smaller matrix C.
+        #Alternatively, U, lambda_matrix = la.eigh(C), but this returns a dense lambda_matrix matrix.
```

Comment:
Calling ``val1, val2 = la.eigh(C)`` puts the eigenvalues in ``val1`` and eigenvectors in ``alg2``.

Reply:
``eigh`` should be used instead of ``svd``